tests/Integration/Behaviour/Features/Context/Domain/SqlManagerFeatureContext.php

```diff
@@ -0,0 +1,129 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Integration\Behaviour\Features\Context\Domain;
+
+use Behat\Gherkin\Node\TableNode;
+use Exception;
+use PHPUnit\Framework\Assert;
+use PrestaShop\PrestaShop\Core\Domain\SqlManagement\Exception\SqlRequestConstraintException;
+use RuntimeException;
+use Symfony\Component\HttpFoundation\Request;
+use Tests\Integration\Behaviour\Features\Context\SharedStorage;
+
+class SqlManagerFeatureContext extends AbstractDomainFeatureContext
+{
+    private const SQL_REQUEST_FORM_KEY = 'sql-request-form';
+
+    /**
+     * @Given I specify following properties for sql request form
+     */
+    public function specifyPropertiesForSqlRequestForm(TableNode $node): void
+    {
+        $data = $node->getRowsHash();
+
+        SharedStorage::getStorage()->set(static::SQL_REQUEST_FORM_KEY, $data);
+    }
+
+    /**
+     * @When I submit the sql request form
+     */
+    public function submitTheSqlRequestForm(): void
```

Comment:
I do not understand what you do in this test.

You modified the behavior of the class SqlRequestFormDataHandler. The class SqlRequestFormDataHandler has a very simple API. Why not test the class SqlRequestFormDataHandler?

Why write a test that involves services `prestashop.core.form.builder.sql_request_form_builder` and `prestashop.core.form.identifiable_object.sql_request_form_handler` even though they are not concerned?

Morever, service `prestashop.core.form.builder.sql_request_form_builder` is a FormBuilder service that belongs to the application layer. In order to test it, you have to fake creating a request using ` $request = Request::createFromGlobals();`. I think it indicates this test does not belong to this layer.

Our Behat tests are supposed to test _behaviors_ of our code, our domain logic. Our domain logic should - as much as possible - be unaware whether behaviors are done through a web interface or a CLI.

I think if we write some Behat tests that depend on a service which has "form" in its name, we are doing something wrong. I am sorry but I think this is not the right test layer.

Reply:
> You modified the behavior of the class SqlRequestFormDataHandler. The class SqlRequestFormDataHandler has a very simple API. Why not test the class SqlRequestFormDataHandler?

I want to test the whole process and validation.

> Why write a test that involves services prestashop.core.form.builder.sql_request_form_builder and prestashop.core.form.identifiable_object.sql_request_form_handler even though they are not concerned?

It's an integration test, I am testing the whole process, as it is done in the Core.

> Morever, service prestashop.core.form.builder.sql_request_form_builder is a FormBuilder service that belongs to the application layer. In order to test it, you have to fake creating a request using $request = Request::createFromGlobals();. I think it indicates this test does not belong to this layer.

I see no problem creating fake request for purpose :thinking: 

> Our Behat tests are supposed to test behaviors of our code, our domain logic. Our domain logic should - as much as possible - be unaware whether behaviors are done through a web interface or a CLI.

Behat is a tool that makes behavior driven development (BDD) possible with human-readable stories that describe the behavior of the application. This is exactly what I'm doing here :thinking:

Reply:
Let's cut long discussions and ask @eternoendless his opinion about this test

Reply:
According to what @matks said, If we want make it testable as we used to do it, the SQL request validation have to be performed in the AddSqlRequestHandler and EditSqlRequestHandler
In this way we will be able to test without the formHandler and the whole request

Reply:
We have several kinds of tests, and the frontier between them is not self-evident, so it's understandable that there might be confusion and a diversity of opinions.

To me it all comes down to what are you trying to test:

- Are you testing what the user sees? While Behat could be an adequate tool for that, we use UI tests to test that.
- Are you testing the domain behavior? Go for Behat.
- Are you testing several components together? Integration (phpunit)
- Are you testing a single component? Unit

Remember the domain enforces validation so that people cannot do with it what they're not supposed to. Forms on the other hand are meant to provide a user interface to the domain, and while they do perform data validation, it is there to provide feedback about what exactly the user did wrong, not as a barrier to protect the system (that is the domain's job).

In conclusion, as @sowbiba said, I think your behavior tests should be checking domain exceptions thrown by `AddSqlRequestHandler` and `EditSqlRequestHandler`, not the form handler. If you want to test the actual error messages sent by the form handler, it would be more appropriate to test that with a UI test.

Both approaches will lead to test cases that are simpler, more robust, and easier to maintain.